### PR TITLE
Support CRAM reference regions.

### PR DESCRIPTION
**Pull request by cmnbroad**
Fixes https://github.com/samtools/htsjdk/issues/1597 (support IGV browse more for CRAM decoding).

### CODE REVIEW

File: src/main/java/htsjdk/samtools/cram/build/SliceFactory.java [hunk 1]
```diff
@@ -136,7 +136,13 @@ public List<Slice> createSlices(
                     containerByteOffset,
                     sliceStagingEntry.getGlobalRecordCounter()
             );
-            slice.setReferenceMD5(cramReferenceRegion.getCurrentReferenceBases());
+            if (slice.getAlignmentContext().getReferenceContext().isMappedSingleRef()) {
+                cramReferenceRegion.getReferenceBasesByRegion(
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/cram/build/SliceFactory.java:140]: It's a bit weird to use the getter to update the internal state in order to make the call to "getCurrentBases" work.  As far as I see it's actually only ever used for it's side effect and the return value is always ignored except in tests.  Should the api change to make it clear that that's the primary use?  Should the uses change to use the return value?  Unclear... Maybe add a default method `setCurrentRegion()` that just calls get and throws away the result, just to make it clear that this is not a no-op.

**Review by lbergelson — CHANGES_REQUESTED** [review]
@cmnbroad I think this looks good overall but I have a bunch of comments of varying levels of usefulness.  I assume this is assuming there's an IGV specific CramReferenceSource which is able to load chunks?  

It's a bit weird that we don't actually use the return value from the new method.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 2]
```diff
@@ -25,72 +25,200 @@
...
+     * @return the currently cached reference bases (may return null)
      */
     public byte[] getCurrentReferenceBases() {
         return referenceBases;
     }
 
     /**
-     * Return the reference bases for the given reference index.
-     * @param referenceIndex
+     * @return the current reference index ({@link ReferenceContext#UNINITIALIZED_REFERENCE_ID} if no index has
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:75]: @lbergelson Can you rephrase this - I'm not sure what you mean by separator ? I did notice a missing closing paren, which I fixed.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 3]
```diff
@@ -25,72 +25,200 @@
...
+    public int getRegionLength() {
+        return regionLength;
+    }
+
+    /**
+     * Return the reference bases for an entire contig given a reference contig index.
      *
-     * @return bases for the entire reference contig specifed by {@code referenceIndex}
+     * @param referenceIndex the reference index for which bases should be retrieved
+     * @return bases for the entire reference contig specified by {@code referenceIndex}, may be null if
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:102]: It can return null in it's initial state if no reference bases have ever  been established.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 4]
```diff
@@ -25,72 +25,200 @@
...
-                    throw new IllegalArgumentException(
-                            String.format(
-                                    "A reference must be supplied (reference sequence %s not found).",
-                                    sequence));
-                }
-
-                referenceBasesContextID = referenceIndex;
+        ValidationUtils.validateArg(referenceIndex >= 0, "reference index must be >= 0");
+
+        // Re-resolve our the reference bases if we don't have a current region or the one we have
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:109]: Fixed.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 5]
```diff
@@ -25,72 +25,200 @@
...
+        // doesn't span the *entire* contig requested. When called on behalf of non-coord sorted inputs,
+        // this can cause a lot of thrashing.
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:111]: Yeah, done.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 6]
```diff
@@ -25,72 +25,200 @@
...
 public class CRAMReferenceRegion {
+    private static final Log log = Log.getInstance(CRAMReferenceRegion.class);
+    public static final int UNINITIALIZED_OFFSET = -1;
+    public static final int  UNINITIALIZED_LENGTH = -1;
 
     private final CRAMReferenceSource referenceSource;
     private final SAMFileHeader samFileHeader;
 
-    private byte[] referenceBases = null; // cache the reference bases
     private int referenceBasesContextID = ReferenceContext.UNINITIALIZED_REFERENCE_ID;
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:49]: Yeah, embedded or not they're all from the same namespace of actual reference index values. Fixed.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 7]
```diff
@@ -25,72 +25,200 @@
...
+        // Re-resolve our the reference bases if we don't have a current region or the one we have
+        // doesn't span the *entire* contig requested. When called on behalf of non-coord sorted inputs,
+        // this can cause a lot of thrashing.
+        if ((referenceIndex != referenceBasesContextID) ||
+                regionOffset != 0 ||
+                (regionLength != referenceBases.length)) {
+            sequenceRecord = getSAMSequenceRecord(referenceIndex);
+            if (sequenceRecord == null) {
+                throw new IllegalArgumentException(
+                        String.format("A reference must be supplied (reference sequence %s not found).", sequenceRecord));
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:118]: Yeah, this is removed entirely.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 8]
```diff
@@ -25,72 +25,200 @@
...
         }
 
-        // retain whatever cached reference bases we may have to minimize subsequent re-fetching
-        return null;
+        if (referenceIndex != referenceBasesContextID) {
+            sequenceRecord = getSAMSequenceRecord(referenceIndex);
+        }
+
+        if ((zeroBasedOffset + requestedFragmentLength) > sequenceRecord.getSequenceLength()) {
+            log.warn("Attempt to retrieve a reference fragment (start: " + zeroBasedOffset + " length: "
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:168]: I had been thinking of this as an anomalous case, but I guess it doesn't need to be logged since there is a legitimate case where its a "normal" operation, here and in reference source. Removed in both places.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 9]
```diff
@@ -25,72 +25,200 @@
...
+                    + requestedFragmentLength + ") that extends beyond the end of the reference contig length: "
+                    + sequenceRecord.getSequenceLength());
+        }
+
+        referenceBases = referenceSource.getReferenceBasesByRegion(sequenceRecord, zeroBasedOffset, requestedFragmentLength);
+        if (referenceBases == null) {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:174]: Note to self; add a test for this new condition.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 10]
```diff
@@ -25,72 +25,200 @@
...
+     */
+    public int getFullContigLength() {
+        return sequenceRecord == null ? -1 : sequenceRecord.getSequenceLength();
+    }
+
+    private SAMSequenceRecord getSAMSequenceRecord(final int referenceIndex) {
+        final SAMSequenceRecord samSequenceRecord = samFileHeader.getSequence(referenceIndex);
+        if (samSequenceRecord == null) {
+            throw new IllegalArgumentException(
+                    String.format("A reference must be supplied (reference sequence index %d not found)", referenceIndex));
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:219]: Done.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 11]
```diff
@@ -25,72 +25,200 @@
...
+ * from the previous request, along with metadata about the (0-based) position/offset, and length of the
+ * cached bases.
  */
 public class CRAMReferenceRegion {
+    private static final Log log = Log.getInstance(CRAMReferenceRegion.class);
+    public static final int UNINITIALIZED_OFFSET = -1;
+    public static final int  UNINITIALIZED_LENGTH = -1;
 
     private final CRAMReferenceSource referenceSource;
     private final SAMFileHeader samFileHeader;
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:47]: Hm. Yeah. Done.

File: src/main/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java [hunk 12]
```diff
@@ -8,15 +8,27 @@
...
+     * Get the reference bases for an entire reference contig.
+     *
+     * @param sequenceRecord the SAMSequenceRecord identifying the reference being requested
+     * @param tryNameVariants if true, attempt to match the requested sequence name against the reference by
+     *                       using common name variations, such as adding or removing a leading "chr" prefix
+     *                       from the requested name. if false, use exact match
+     * @return the upper-cased, normalized (see {@link htsjdk.samtools.cram.build.Utils#normalizeBase})
      * bases representing the requested sequence, or null if the sequence cannot be found
      */
     byte[] getReferenceBases(final SAMSequenceRecord sequenceRecord, final boolean tryNameVariants);
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java:20]: I have no idea what the motivation for it originally was (AFAIK it has always been in the code), but I'd love to take it out. It seems like a total hack and I don't see how any good can come of it.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 13]
```diff
@@ -25,72 +25,200 @@
...
+     *
+     * @param embeddedReferenceBases the embedded reference bases to be used
+     * @param embeddedReferenceIndex the reference ID used in the slice containing the embedded reference
+     * @param zeroBasedOffset the zero based reference offset of the first base in the embedded reference bases
+     */
+    public void setEmbeddedReferenceBases(
+            final byte[] embeddedReferenceBases,
+            final int embeddedReferenceIndex,
+            final int zeroBasedOffset) {
+        sequenceRecord = getSAMSequenceRecord(embeddedReferenceIndex);
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:199]: There is almost no detail in the spec about embedded references, but AFAIK these come from the same set of values. I think using any other set of values would be meaningless (a slice that uses an embedded reference still has to have a reference sequence ID in it's header). Also, the one samtools-generated test file that we use for testing embedded references to have a valid value.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 14]
```diff
@@ -25,72 +25,200 @@
...
+    public static final int  UNINITIALIZED_LENGTH = -1;
 
     private final CRAMReferenceSource referenceSource;
     private final SAMFileHeader samFileHeader;
 
-    private byte[] referenceBases = null; // cache the reference bases
     private int referenceBasesContextID = ReferenceContext.UNINITIALIZED_REFERENCE_ID;
+    private byte[] referenceBases = null;
+    private SAMSequenceRecord sequenceRecord = null;
+    private int regionOffset = UNINITIALIZED_OFFSET; // 0-based start offset of the region
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:52]: "start" may be a better choice since it more naturally aligns with "alignment start" in the rest of the code. Changed.

File: src/main/java/htsjdk/samtools/cram/ref/ReferenceSource.java [hunk 15]
```diff
@@ -179,6 +187,46 @@ public synchronized byte[] getReferenceBases(final SAMSequenceRecord record,
...
+                    getReferenceBases(sequenceRecord, false) :
+                    currentBackingBases;
+
+        if (bases != null) {
+            // cache the backing bases to prevent thrashing due to aggressive GC
+            backingReferenceBases = bases;
+            backingContigIndex = sequenceRecord.getSequenceIndex();
+
+            if ((zeroBasedOffset + requestedRegionLength) > bases.length) {
+                log.warn("Can't supply reference bytes for start: " + zeroBasedOffset + " length: " +
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/ref/ReferenceSource.java:211]: Fixed.

File: src/main/java/htsjdk/samtools/cram/ref/ReferenceSource.java [hunk 16]
```diff
@@ -179,6 +187,46 @@ public synchronized byte[] getReferenceBases(final SAMSequenceRecord record,
...
+                        requestedRegionLength + " from sequence: " + sequenceRecord.getSequenceName() +
+                                " of length " + bases.length + " bytes");
+                final byte[] basesSlice = Arrays.copyOfRange(bases, zeroBasedOffset, bases.length);
+                return basesSlice;
+            } else {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/ref/ReferenceSource.java:216]: I pulled out the `copyOfRange` call (and got rid of the intermediate variable). There is only one `end` reference now so its calculated inline.

File: src/main/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java [hunk 17]
```diff
@@ -8,15 +8,27 @@
...
+
+    /**
+     * Get the reference bases for a region of a reference contig.
+     *
+     * @param sequenceRecord the SAMSequenceRecord for the reference contig being requested
+     * @param zeroBasedOffset the zero based offset of the starting reference base
+     * @param requestedRegionLength the length of the requested reference region
+     * @return the bases for the reference region
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java:28]: Yes, and also I updated the doc to state that it is permissible to return fewer bases than requested if the requested length exceeds the actual sequence length.

File: src/main/java/htsjdk/samtools/cram/structure/CRAMRecordReadFeatures.java [hunk 18]
```diff
@@ -356,22 +357,24 @@ public Cigar getCigarForReadFeatures(final int readLength) {
...
-     * @param substitutionMatrix
-     * @return
+     * Get the read bases for a CRAMRecord given a set of read feaures and a reference region.
+     *
+     * @param readFeatures list of ReadFeatures for this record. may be null
+     * @param isUnknownBases true if CF_UNKNOWN_BASES CRAM flag is set for this read
+     * @param readAlignmentStart 1-based CRAM record alignment start
+     * @param readLength read length for this read
+     * @param cramReferenceRegion CRAMReferenceRegion spanning the reference bases for this read,
+     *                           if reference-compressed. the region must already be resolved to the correct
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/structure/CRAMRecordReadFeatures.java:367]: I meant that this method does uses the whatever the current CRAMReferenceRegion state is (the current bases in the CRAMReferenceRegion), and does not attempt to fetch them first. Essentially, it's saying that it's the caller's responsibility to have already resolved/fetched the correct bases. I'll try to clarify that.

File: src/test/java/htsjdk/samtools/cram/ref/CRAMReferenceRegionTest.java [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+                {
+                        // entirety of contig 1
+                        getAlternatingReferenceRegion(),
+                        CRAMStructureTestHelper.REFERENCE_SEQUENCE_ONE,
+                        0,
+                        CRAMStructureTestHelper.REFERENCE_CONTIG_LENGTH,
+                        true
+                },
+                {
+                        // 1first byte of contig 0
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/test/java/htsjdk/samtools/cram/ref/CRAMReferenceRegionTest.java:85]: fixed.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java [hunk 20]
```diff
@@ -25,72 +25,200 @@
...
-    public void setEmbeddedReference(final byte[] embeddedReferenceBytes, final int embeddedReferenceIndex) {
+    /**
+     * Set this {@link CRAMReferenceRegion} to use an embedded reference.
+     *
+     * @param embeddedReferenceBases the embedded reference bases to be used
+     * @param embeddedReferenceIndex the reference ID used in the slice containing the embedded reference
+     * @param zeroBasedOffset the zero based reference offset of the first base in the embedded reference bases
+     */
+    public void setEmbeddedReferenceBases(
+            final byte[] embeddedReferenceBases,
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:196]: Yeah done.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:49]: Sure, done.

File: src/main/java/htsjdk/samtools/cram/build/SliceFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/SliceFactory.java:140]: Done.

Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/SliceFactory.java:140]: Yeah its a little wonky, and its true there are no longer any (non-test) call sites that use the return value directly. I updated the methods to have a `void` return type and renamed them with a `fetch` prefix instead of a get prefix to reinforce their impure character. Let me know if you think thats sufficient.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:77]: Oh, I meant that you need a comma or linebreak between `reference index `and `{{@link`` so it's clear those are 2 different cases.

File: src/main/java/htsjdk/samtools/cram/build/SliceFactory.java (hunk 1 above)
Comment (lbergelson) [reply src/main/java/htsjdk/samtools/cram/build/SliceFactory.java]: That sounds good to me.

File: src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/htsjdk/samtools/cram/build/CRAMReferenceRegion.java:170]: That's reasonable. 👍

**Comment by cmnbroad** [comment]
Ok @lbergelson, back to you.

**Review by lbergelson — APPROVED** [review]
@cmnbroad This looks good to me.  One potential follow up...  maybe we should make `getReferenceBases(contig)` into a default method that is implemented in terms of `getReferenceBasesByRegion`.  

It's not totally clear we actually want to keep them as 0-based numbering since we convert in in all of our calls.  I'm going to merge this as is and we can think about changes later.
